File: src/main/java/cn/solarmoon/idyllic_food_diary/feature/generic_recipe/stir_fry/StirFryStage.java
package cn.solarmoon.idyllic_food_diary.feature.generic_recipe.stir_fry;

import cn.solarmoon.idyllic_food_diary.feature.spice.SpiceList;
import cn.solarmoon.solarmoon_core.api.data.SerializeHelper;
import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.util.GsonHelper;
import net.minecraft.world.item.crafting.Ingredient;
import net.minecraftforge.fluids.FluidStack;

import java.util.ArrayList;
import java.util.List;

public record StirFryStage(
        List<Ingredient> ingredients,
        FluidStack fluidStack,
        SpiceList spices,
        int time,
        int fryCount,
        boolean keepFluid
) {

    public static StirFryStage readFromJson(JsonObject json) {
        List<Ingredient> ingredients = SerializeHelper.readIngredients(json, "ingredients");
        FluidStack fluidStack = SerializeHelper.readFluidStack(json, "fluid");
        SpiceList spices = SpiceList.readSpices(json, "with_spices");
        int time = GsonHelper.getAsInt(json, "time", 0);
        int fryCount = GsonHelper.getAsInt(json, "fry_count", 0);
        boolean keepFluid = GsonHelper.getAsBoolean(json, "keep_fluid", true);
        return new StirFryStage(ingredients, fluidStack, spices, time, fryCount, keepFluid);
    }

    public static StirFryStage readFromNetwork(FriendlyByteBuf buf) {
        List<Ingredient> ingredients = SerializeHelper.readIngredients(buf);
        FluidStack fluidStack = buf.readFluidStack();
        SpiceList spices = SpiceList.readSpices(buf);
        int time = buf.readInt();
        int fryCount = buf.readInt();
        boolean keepFluid = buf.readBoolean();
        return new StirFryStage(ingredients, fluidStack, spices, time, fryCount, keepFluid);
    }

    public void toNetwork(FriendlyByteBuf buf) {
        SerializeHelper.writeIngredients(buf, ingredients);
        buf.writeFluidStack(fluidStack);
        SpiceList.writeSpices(buf, spices);
        buf.writeInt(time);
        buf.writeInt(fryCount);
        buf.writeBoolean(keepFluid);
    }

    public static List<StirFryStage> readListFromJson(JsonObject json, String id) {
        if (json.has(id)) {
            List<StirFryStage> stirFryStages = new ArrayList<>();
            JsonArray jsonArray = GsonHelper.getAsJsonArray(json, "stir_fry_stages");
            for (var element : jsonArray) {
                stirFryStages.add(readFromJson(element.getAsJsonObject()));
            }
            return stirFryStages;
        }
        return new ArrayList<>();
    }

    public static List<StirFryStage> readListFromNetwork(FriendlyByteBuf buf) {
        int size = buf.readVarInt();
        List<StirFryStage> stirFryStages = new ArrayList<>();
        for (int i = 0; i < size; i++) {
            stirFryStages.add(readFromNetwork(buf));
        }
        return stirFryStages;
    }

    public static void writeListToNetwork(FriendlyByteBuf buf, List<StirFryStage> stirFryStages) {
        buf.writeVarInt(stirFryStages.size());
        for (var stage : stirFryStages) {
            stage.toNetwork(buf);
        }
    }

}

File: src/main/kotlin/cn/solarmoon/idyllic_food_diary/element/recipe/assistant/StirFryStage.kt
package cn.solarmoon.idyllic_food_diary.element.recipe.assistant

import cn.solarmoon.spark_core.api.data.SerializeHelper
import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder
import net.minecraft.network.RegistryFriendlyByteBuf
import net.minecraft.network.codec.ByteBufCodecs
import net.minecraft.network.codec.StreamCodec
import net.minecraft.world.item.crafting.Ingredient
import net.neoforged.neoforge.fluids.FluidStack

@JvmRecord
data class StirFryStage(
    val ingredients: List<Ingredient>,
    val fluidStack: FluidStack,
    val time: Int,
    val fryCount: Int,
    val keepFluid: Boolean
) {

    companion object {
        @JvmStatic
        val CODEC: Codec<StirFryStage> = RecordCodecBuilder.create { instance ->
            instance.group(
                Ingredient.LIST_CODEC.fieldOf("ingredients").forGetter { it.ingredients },
                FluidStack.OPTIONAL_CODEC.fieldOf("input_fluid").forGetter { it.fluidStack },
                Codec.INT.fieldOf("time").forGetter { it.time },
                Codec.INT.fieldOf("fry_count").forGetter { it.fryCount },
                Codec.BOOL.optionalFieldOf("keep_fluid", true).forGetter { it.keepFluid }
            ).apply(instance, ::StirFryStage)
        }

        @JvmStatic
        val LIST_CODEC = CODEC.listOf()

        @JvmStatic
        val STREAM_CODEC = object : StreamCodec<RegistryFriendlyByteBuf, StirFryStage> {
            override fun decode(buffer: RegistryFriendlyByteBuf): StirFryStage {
                val ingredients = SerializeHelper.INGREDIENT.LIST_STREAM_CODEC.decode(buffer)
                val inputFluid = FluidStack.OPTIONAL_STREAM_CODEC.decode(buffer)
                val time = buffer.readInt()
                val fryCount = buffer.readInt()
                val keepFluid = buffer.readBoolean()
                return StirFryStage(ingredients, inputFluid, time, fryCount, keepFluid)
            }

            override fun encode(buffer: RegistryFriendlyByteBuf, stage: StirFryStage) {
                SerializeHelper.INGREDIENT.LIST_STREAM_CODEC.encode(buffer, stage.ingredients)
                FluidStack.OPTIONAL_STREAM_CODEC.encode(buffer, stage.fluidStack)
                buffer.writeInt(stage.time)
                buffer.writeInt(stage.fryCount)
                buffer.writeBoolean(stage.keepFluid)
            }
        }

        @JvmStatic
        val LIST_STREAM_CODEC: StreamCodec<RegistryFriendlyByteBuf, List<StirFryStage>> = STREAM_CODEC.apply(ByteBufCodecs.collection { mutableListOf() })
    }

}